### PR TITLE
Insert text via API

### PR DESCRIPTION
**Pull request by c-martinez**
Implements feature suggested on issue #2897.

Not quite yet finished, but would like to check this is going in the right direction. Feedback welcome!

### CODE REVIEW

File: src/node/handler/APIHandler.js [hunk 1]
```diff
@@ -56,6 +56,7 @@ var version =
   , "listSessionsOfAuthor"      : ["authorID"]
   , "getText"                   : ["padID", "rev"]
   , "setText"                   : ["padID", "text"]
+  , "insertText"                : ["padID", "text", "lineNum"]
```
Comment (Gared) [inline src/node/handler/APIHandler.js]: Add your new API method only to the latest API version

**Comment by c-martinez** [comment]
Ah, yes -- I started by adding it to every version while I was figuring it out. Removed.

**Comment by JohnMcLear** [comment]
This needs backend tests plz :)

**Comment by c-martinez** [comment]
Better?

**Comment by Gared** [comment]
Looks like the pad is being destroyed if you remove the "lineNumber" parameter (I get javascript errors after calling the API without this parameter). Please make this parameter necessary.

**Comment by Gared** [comment]
Also please set up a "real author" otherwise you'll get the error message "There is no author for authorId: API-bot"

**Comment by JohnMcLear** [comment]
Thanks for test coverage @c-martinez -- Please also add coverage for the bits @Gared has discovered

**Comment by c-martinez** [comment]
@Gared, thanks for feedback. lineNum now is a required parameter.

I'm adding authorID also as a required parameter -- but I'm not sure how to test this -- `createAuthor` doesn't seem to have tests either, and I'm not sure how to get a authorID to use in the tests. Any tips?

**Comment by c-martinez** [comment]
Hi -- added a "real author" (via authorMapper -- not sure if this is the most logical way), and added tests for it as well.

Along the way, added tests for `createAuthor` and `createAuthorIfNotExistsFor`.

comments?

File: src/node/db/API.js [hunk 2]
```diff
@@ -298,15 +298,91 @@ exports.setText = function(padID, text, callback)
...
+  var authorMangager = require("./AuthorManager");
+  // Check author exists
+  authorMangager.createAuthorIfNotExistsFor(authorMapper, false, function(err, authorID)
+  {
+    if(ERR(err, callback)) return;
+
+    //author does not exist
+    if(authorID == false)
+    {
+      callback(new customError("Author not given or given author does not exits","apierror"));
```
Comment (Gared) [inline src/node/db/API.js]: propably there is a misspelling in "exits". Maybe "exist"? :-)

File: src/node/db/API.js [hunk 3]
```diff
@@ -298,15 +298,91 @@ exports.setText = function(padID, text, callback)
...
+        var Changeset = require("ep_etherpad-lite/static/js/Changeset");
+        var oldText = pad.text();
+        // Index of insert point
+        var oldLines = oldText.split("\n");
+        var insertPoint = 0;
+        if(lineNum < oldLines.length) {
+          for (var i = 0; i < lineNum-1; i++) {
+            insertPoint += oldLines[i].length + 1;
+          }
+        } else {
```
Comment (Gared) [inline src/node/db/API.js]: Please use a separate line for the brackets and "else" (code formatting)

**Comment by Gared** [comment]
Looks good to me so far. Thx for your work :-)
Will do some tests the next days.

**Comment by c-martinez** [comment]
Well spotted @Gared , thanks!

File: src/node/db/API.js [hunk 4]
```diff
@@ -298,15 +298,93 @@ exports.setText = function(padID, text, callback)
...
+        var oldLines = oldText.split("\n");
+        var insertPoint = 0;
+        if(lineNum < oldLines.length) {
+          for (var i = 0; i < lineNum-1; i++) {
+            insertPoint += oldLines[i].length + 1;
+          }
+        }
+        else
+        {
+          insertPoint = oldLen;
```
Comment (Gared) [inline src/node/db/API.js]: Where is oldLen set? Seems to me that this variable is not defined at this point

File: src/node/db/API.js [hunk 5]
```diff
@@ -298,15 +298,93 @@ exports.setText = function(padID, text, callback)
...
+  }
+
+  var authorMangager = require("./AuthorManager");
+  // Check author exists
+  authorMangager.createAuthorIfNotExistsFor(authorMapper, false, function(err, authorID)
+  {
+    if(ERR(err, callback)) return;
+
+    //author does not exist
+    if(authorID == false)
```
Comment (Gared) [inline src/node/db/API.js]: Why do you check the authorID? As far as I can see there is no case that the authorID can be 'false'

File: src/node/db/API.js [hunk 6]
```diff
@@ -298,15 +298,93 @@ exports.setText = function(padID, text, callback)
...
+{
+  //text is required
+  if(typeof newText != "string")
+  {
+    callback(new customError("text is no string","apierror"));
+    return;
+  }
+
+  //check if lineNum is a number
+  if(!isNaN(parseInt(lineNum)))
```
Comment (Gared) [inline src/node/db/API.js]: Please do call "parseInt" only once

File: src/node/db/API.js (hunk 4 above)
Comment (Gared) [inline src/node/db/API.js]: This seems to be the reason why you get an error if you specify a lineNum that is higher than the actual number of lines on this pad

**Comment by Gared** [comment]
Please also documentate your new API method here:
https://github.com/ether/etherpad-lite/blob/9335ae00d0c889bf96482447990968ca3eef6806/doc/api/http_api.md

**Comment by c-martinez** [comment]
Hi @Gared , thanks for comments. (Sorry, it looks like you spent quite a bit of time fixing issues which I should have seen before -- really thanks for that). I'll address them and send a new commit.

File: src/node/db/API.js (hunk 5 above)
Comment (c-martinez) [inline src/node/db/API.js]: Good point

File: src/node/db/API.js [hunk 7]
```diff
@@ -298,15 +298,82 @@ exports.setText = function(padID, text, callback)
...
+
+  if(typeof authorMapper != "string")
+  {
+    callback(new customError("Invalid authorMapper","apierror"));
+    return;
+  }
+
+  var authorMangager = require("./AuthorManager");
+  // Check author exists
+  authorMangager.createAuthorIfNotExistsFor(authorMapper, false, function(err, authorID)
```
Comment (Gared) [inline src/node/db/API.js]: In the callback you get the author object and not the authorID

File: src/node/db/API.js [hunk 8]
```diff
@@ -298,15 +298,82 @@ exports.setText = function(padID, text, callback)
...
+      {
+        insertPoint = oldLen-1;
+      }
+
+      var assem = Changeset.smartOpAssembler();
+      assem.appendOpWithText('=', oldText.substring(0, insertPoint));
+      assem.appendOpWithText('+', newText);
+      assem.endDocument();
+      var typedChanges = Changeset.pack(oldLen, newLen, assem.toString(), newText);
+      pad.appendRevision(typedChanges, authorID);
```
Comment (Gared) [inline src/node/db/API.js]: Use here author.authorID

File: src/node/db/API.js (hunk 7 above)
Comment (Gared) [inline src/node/db/API.js]: Is there any special reason why you set the parameter for "name" to "false"?

**Comment by Gared** [comment]
@c-martinez Can you do these changes, please? I want to release a new version of etherpad with this feature.

**Comment by c-martinez** [comment]
Hi @Gared -- yes, sorry, I've been distracted with other things. I'll make these changes later today.

**Comment by c-martinez** [comment]
Done!

But I get the impression that I'm still doing something not-quite-right... I'm using `createAuthorIfNotExistsFor` but I don't think I should be creating a new author. Is there a checkIfAuthorExistsForMapper function (or something similar?)

**Comment by Gared** [comment]
@c-martinez No, there is currently no such function, but I'm fine to use the author mapper to create an author if necessary. So I would merge your PR now.

**Comment by c-martinez** [comment]
Would it be a potentially useful feature?

**Comment by Gared** [comment]
@c-martinez I think you're right. If we use the authorMapper to create a new author there is no way to insert text as another author or create an author with a name. So I would prefer to replace the parameter ```authorMapper``` with ```authorID```. There should be a check if this authorID exists, but nothing more. What do you think about?

To answer your question: I don't think such a function would be useful, because the authorMapper is a random string created by an application using the API. In my opinion there is no reason to ask etherpad for an authorMapper.

**Comment by c-martinez** [comment]
I guess using an authorID makes more sense then.

**Comment by Gared** [comment]
@c-martinez Thanks for your quick changes, but I've noticed that there is a problem after applying a text to a pad:
1. Open a pad
2. Write something in the pad
3. Call insertText-API with the authorID of your current user in the Pad
4. Write something in the pad again
5. You get a disconnect in the pad
```
[WARN] client - Error: applySubmittedChangesToBase: no submitted changes to apply -- { errorId: 'pp8WgPgOyFStijvP4rXA',
  msg: 'Error: applySubmittedChangesToBase: no submitted changes to apply',
  url: 'http://localhost:9001/p/test',
  linenumber: 10465,
  userAgent: 'Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:45.0) Gecko/20100101 Firefox/45.0' }
```

It seems that there is a problem with the author and the change being applied to the pad

**Comment by c-martinez** [comment]
@Gared -- I'll look into it.

**Comment by c-martinez** [comment]
@Gared -- reproduced the issue you mentioned. It seems like the API call itself is working fine, but the browser session gets disconnected or something.

We could avoid this by not using a normal author to insert text via the API (i.e. do not have an authorID as a parameter). Instead we could create a 'special' author (via `createAuthorIfNotExistsFor` for instance), and always use the same author for API calls. This should solve the issue, right? What do you think?

**Comment by Gared** [comment]
@c-martinez That would work, but I would really appreciate if we can add text as any author via the API.
I think there is a problem with the authorID in the change, because the added text via the API isn't colored and has no author attributes. This looks like the main problem to me.

**Comment by fcassin** [comment]
The problem seems to be that the client receives an ACCEPT_COMMIT message but there was no submittedChangeset created beforehand. This is what causes the uncaught error.

If you have two users connected on the same pad when the API call is made, the user not "impersonated" by the API sees the changes immediatly, but with no author color. Only the impersonated user loses its session after the API call. Maybe a NEW_CHANGES message is required, instead of a ACCEPT_COMMIT message.

**Comment by c-martinez** [comment]
Following @Gared suggestion on #2950, this should add the colour to the text according to the colour set by the user. (Maybe need to check if author already has attributes set for this pad).

**Comment by JohnMcLear** [comment]
@c-martinez is this something you still need?

**Comment by c-martinez** [comment]
Not really -- I am not using etherpad at the moment.

**Pull request closed without merging**